test_util/cluster_api.py

```diff
@@ -263,39 +276,76 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
 
         assert len(self.masters) == len(self.public_masters)
 
-        # URI must include scheme
-        assert dcos_uri.startswith('http')
-        parse_result = urlparse(dcos_uri)
-        self.scheme = parse_result.scheme
-        self.dns_host = parse_result.netloc.split(':')[0]
-
-        # Make URI never end with /
-        self.dcos_uri = dcos_uri.rstrip('/')
-
     def get_user_session(self, user):
-        """ Return a copy of self instead of new instance of ClusterApi
-        so that children of this class return siblings, not parents
+        """Returns a copy of self with auth headers set for user
         """
         new_session = copy.deepcopy(self)
-        if user:
-            user.authenticate(self)
+        # purge old auth headers
+        if self.web_auth_default_user is not None:
+            for k in self.web_auth_default_user.auth_header.keys():
+                del new_session.default_headers[k]
+        # if user is given then auth and update the headers
+        if user is not None:
+            new_session._authenticate_default_user()
         new_session.web_auth_default_user = user
         return new_session
 
-    def get_client(self, path, default_headers=None):
-        return test_util.helpers.ApiClient(
-            cluster=self,
-            user=self.web_auth_default_user,
+    def get_node_url(self, node, port=None):
+        """
+        Args:
+            node: (str) the hostname of the node to be requested from, if node=None,
+                then public cluster address (see environment DCOS_DNS_ADDRESS)
+            port: (int) port to be requested at. If port=None, the default port
+                for that given node type will be used
+        Returns:
+            fully-qualified URL string for this API
+        """
+        if node in self.masters:
+            role = 'master'
+        elif node in self.all_slaves:
+            role = 'agent'
+        else:
+            raise Exception('Node {} is not recognized within the DC/OS cluster'.format(node))
```

Comment:
Which specific one would you prefer? I prefer generic Exceptions with specific messages until we actually are trying to catch the exception.